mono/utils/mono-mmap-windows.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * mono-mmap-windows.c: Windows support for mapping code into the process address space
+ *
+ * Author:
+ *   Mono Team (mono-list@lists.ximian.com)
+ *
+ * Copyright 2001-2008 Novell, Inc.
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+#include <config.h>
+#include <glib.h>
+
+#if defined(HOST_WIN32)
+#include <Windows.h>
+#include "mono/utils/mono-mmap-windows.h"
+#include <mono/utils/mono-counters.h>
+#include <io.h>
+
+static void *malloced_shared_area = NULL;
+
+int
+mono_pagesize (void)
+{
+	SYSTEM_INFO info;
+	static int saved_pagesize = 0;
+	if (saved_pagesize)
+		return saved_pagesize;
+	GetSystemInfo (&info);
+	saved_pagesize = info.dwAllocationGranularity;
+	return saved_pagesize;
+}
+
+int
+mono_mmap_win_prot_from_flags (int flags)
+{
+	int prot = flags & (MONO_MMAP_READ|MONO_MMAP_WRITE|MONO_MMAP_EXEC);
+	switch (prot) {
+	case 0: prot = PAGE_NOACCESS; break;
+	case MONO_MMAP_READ: prot = PAGE_READONLY; break;
+	case MONO_MMAP_READ|MONO_MMAP_EXEC: prot = PAGE_EXECUTE_READ; break;
+	case MONO_MMAP_READ|MONO_MMAP_WRITE: prot = PAGE_READWRITE; break;
+	case MONO_MMAP_READ|MONO_MMAP_WRITE|MONO_MMAP_EXEC: prot = PAGE_EXECUTE_READWRITE; break;
+	case MONO_MMAP_WRITE: prot = PAGE_READWRITE; break;
+	case MONO_MMAP_WRITE|MONO_MMAP_EXEC: prot = PAGE_EXECUTE_READWRITE; break;
+	case MONO_MMAP_EXEC: prot = PAGE_EXECUTE; break;
+	default:
+		g_assert_not_reached ();
+	}
+	return prot;
+}
+
+void*
+mono_valloc (void *addr, size_t length, int flags, MonoMemAccountType type)
+{
+	void *ptr;
+	int mflags = MEM_RESERVE|MEM_COMMIT;
+	int prot = mono_mmap_win_prot_from_flags (flags);
+	/* translate the flags */
+
+	ptr = VirtualAlloc (addr, length, mflags, prot);
+
+	account_mem (type, (ssize_t)length);
+
+	return ptr;
+}
+
+void*
+mono_valloc_aligned (size_t length, size_t alignment, int flags, MonoMemAccountType type)
```

Comment:
Looked a little deeper into the original implementation. First, this code looks like it did when initially implemented back in 2011, commit bcb58869b36687a559ba591f902bd0a31610cf0e. Second, it is not possible on windows to return back parts of what previously been allocated with VirtualAlloc, so it is not possible too free memory before or after the aligned memory block. What you can do, and this is what’s implemented, is to reserve memory with VirtualAlloc and then only commit what’s really needed from aligned address using requested size. You will always take the hit on virtual memory, but you won't consume more memory in page file or physical RAM (plus any page size alignment done by VirtualAlloc).